ambry-named-mysql/src/integration-test/java/com/github/ambry/named/MySqlNamedBlobDbIntegrationTest.java

```diff
@@ -266,6 +266,66 @@ record = new NamedBlobRecord(account.getName(), container.getName(), blobName, b
         namedBlobDb.get(account.getName(), container.getName(), blobName).get());
   }
 
+  /**
+   * Test behavior with list named blob
+   */
+  @Test
+  public void testListNamedBlobsWithStaleRecords() throws Exception {
+    long now = Calendar.getInstance(TimeZone.getTimeZone("UTC")).getTimeInMillis();
+    Account account = accountService.getAllAccounts().iterator().next();
+    Container container = account.getAllContainers().iterator().next();
+    String blobName = "testListNamedBlobsWithStaleRecords";
+    NamedBlobRecord v1, v1_other, v2, v2_other;
+    Page<NamedBlobRecord> page;
+
+    // put blob Ready and list should return the blob
+    v1 = new NamedBlobRecord(account.getName(), container.getName(), blobName, "b1-ready",
+        now + TimeUnit.MINUTES.toMillis(5));
+    v1_other = new NamedBlobRecord(account.getName(), container.getName(), blobName + "-other", "b1-ready-other",
+        now + TimeUnit.MINUTES.toMillis(5));
+    namedBlobDb.put(v1, NamedBlobState.READY, true).get();
+    namedBlobDb.put(v1_other, NamedBlobState.READY, true).get();
+    page = namedBlobDb.list(account.getName(), container.getName(), blobName, null, null).get();
+    assertEquals(2, page.getEntries().size());
+    assertEquals(v1, page.getEntries().get(0));
+    assertEquals(v1_other, page.getEntries().get(1));
+    time.sleep(100);
+
+    // put blob in-progress and list should return the Ready blob
+    v2 = new NamedBlobRecord(account.getName(), container.getName(), blobName, "b2-in-progress",
+        now + TimeUnit.MINUTES.toMillis(5));
+    namedBlobDb.put(v2, NamedBlobState.IN_PROGRESS, true).get();
+    page = namedBlobDb.list(account.getName(), container.getName(), blobName, null, null).get();
+    assertEquals(2, page.getEntries().size());
+    assertEquals(v1, page.getEntries().get(0));
+    assertEquals(v1_other, page.getEntries().get(1));
+    time.sleep(100);
+
+    // update blob and list should return the new blob
+    v2 = new NamedBlobRecord(account.getName(), container.getName(), blobName, "b2-ready",
+        now + TimeUnit.MINUTES.toMillis(5));
+    v2_other = new NamedBlobRecord(account.getName(), container.getName(), blobName + "-other", "b2-ready-other",
+        now + TimeUnit.MINUTES.toMillis(5));
+    namedBlobDb.put(v2, NamedBlobState.READY, true).get();
+    namedBlobDb.put(v2_other, NamedBlobState.READY, true).get();
+    page = namedBlobDb.list(account.getName(), container.getName(), blobName, null, null).get();
+    assertEquals(2, page.getEntries().size());
+    assertEquals(v2, page.getEntries().get(0));
+    assertEquals(v2_other, page.getEntries().get(1));
+    time.sleep(100);
+
+    // delete blob and list should return empty
+    namedBlobDb.delete(account.getName(), container.getName(), blobName).get();
```

Comment:
Can you please add some test like like
1. add one permanent version.
2. upsert the named blob and delete it.
In this case, we should list nothing cause the latest version already been deleted.

Reply:
Yeah, i have added a test